lib/ood_core/job/adapters/htcondor.rb

```diff
@@ -0,0 +1,473 @@
+                    # @return [String,nil] the path to the user/group map file
+                    attr_reader :user_group_map
+
+                    # The cluster name for this HTCondor instance
+                    # @return [String] the cluster name
+                    attr_reader :cluster
+
+                    # Additional attributes to be added to the job submission
+                    # @return [Hash{#to_s => #to_s}] additional attributes to be added to the job submission
+                    attr_reader :additional_attributes
+
+                    # The root exception class that all HTCondor-specific exceptions inherit
+                    # from
+                    class Error < StandardError; end
+
+                    # @param bin [#to_s] path to HTCondor installation binaries
+                    # @param submit_host [#to_s] Submits the job on a login node via ssh
+                    # @param strict_host_checking [Bool] Whether to use strict host checking when ssh to submit_host
+                    def initialize(bin: nil, bin_overrides: {}, submit_host: "", strict_host_checking: false, default_universe: "vanilla", default_docker_image: "ubuntu:latest", user_group_map: nil, cluster: "", additional_attributes: {})
+                        @bin                  = Pathname.new(bin.to_s)
+                        @bin_overrides        = bin_overrides
+                        @submit_host          = submit_host.to_s
+                        @strict_host_checking = strict_host_checking
+                        @default_universe     = default_universe.to_s
+                        @default_docker_image = default_docker_image.to_s
+                        @user_group_map       = user_group_map.to_s unless user_group_map.nil?
+                        @cluster              = cluster.to_s
+                        @additional_attributes = additional_attributes
+                    end
+
+                    # Submit a script to the batch server
+                    # @param args [Array<#to_s>] arguments passed to `condor_submit` command
+                    # @param env [Hash{#to_s => #to_s}] environment variables set
+                    # @param script [String] the script to submit
+                    # @raise [Error] if `condor_submit` command exited unsuccessfully
+                    # @return [String] the id of the job that was created
+                    def submit_string(args: [], script_args: [], env: {}, script: "")
+                        args = args.map(&:to_s)
+                        script_args = script_args.map(&:to_s).map { |s| s.to_s.gsub('"', "'") } # cannot do double 
+                        env = env.to_h.each_with_object({}) { |(k, v), h| h[k.to_s] = v.to_s }
+
+                        tempfile = Tempfile.new("htcondor_submit")
+                        tempfile.close
+                        path = tempfile.path
+                        tempfile.unlink # unlink the tempfile so it can be used by condor_submit
```

Comment:
Instead of creating a Tempfile, can we just generate a unique name? I feel like file operations are expensive, wouldn't `"#{Dir.tmpdir}/#{SecureRandom.uuid}"` or something similar provide the same feature?